lib/autoproj/daemon/github_watcher.rb

```diff
@@ -1,189 +1,224 @@
 # frozen_string_literal: true
 
 require 'autoproj'
-require 'octokit'
-require 'time'
+require 'autoproj/daemon/github/client'
+require 'autoproj/daemon/package_repository'
+require 'autoproj/daemon/github/push_event'
+require 'autoproj/daemon/github/pull_request_event'
 
 module Autoproj
     module Daemon
         # The GithubWatcher will keep an internal list of watched repositories,
         # push hooks and pull request hooks. Whenever an event is detected on a
         # watched repository the corresponding hook is called.
         class GithubWatcher
-            attr_reader :watched_repositories
+            attr_reader :cache
+            attr_reader :client
+            attr_reader :packages
             attr_reader :ws
 
+            # @param [Github::Client] client The github API wrapper
+            # @param [Array<PackageRepository>] packages Watched packages
             # @param [Autoproj::Workspace] ws The loaded workspace
-            def initialize(ws) # rubocop:disable Naming/UncommunicativeMethodParamName
-                @watched_repositories = []
+            # rubocop:disable Naming/UncommunicativeMethodParamName
+            def initialize(client, packages, cache, ws)
+                @cache = cache
+                @client = client
+                @packages = packages
                 @ws = ws
                 @api_key = ws.config.daemon_api_key
                 @polling_period = ws.config.daemon_polling_period
-                @client = Octokit::Client.new(access_token: @api_key)
-                @client.auto_paginate = true
-                @pullrequest_hooks = []
+                @pull_request_hooks = []
                 @push_hooks = []
             end
+            # rubocop:enable Naming/UncommunicativeMethodParamName
 
-            # Attempts to authenticate using the configured API key.
-            # The API does not require us to aunthenticate first but
-            # this is useful so we can get Github's current time.
-            #
-            # @return [Time] The current time provided by GitHub
-            def login
-                @client.user.login
-                Time.parse(@client.last_response.headers[:date])
+            # @return [String] An array with all users that we will be polling
+            def owners
+                @packages.map(&:owner).uniq
             end
 
-            Repository = Struct.new :owner, :name, :branch, :timestamp
+            # @param [Array<Github::PushEvent, Github::PullRequestEvent>] events
+            # @return [(String, String)]
+            def event_owner_and_name(event)
+                if event.kind_of? Github::PushEvent
```

Comment:
dispatching against class like this is usually done with `case ... when`

~~~
case event
when Github::PushEvent
when Github::PullRequestEvent
else
end
~~~